Map visualization

thorncorona opened the pull request:
# Description

- Added `mapviz.html` map visualization
- Mocked backend using DAO pattern.

# Checklist:

- [x] Code follows the style guidelines
- [x] Code commented in relevant places
- [x] Changes to necessary documentation made
- [x] New and existing unit tests pass locally with changes
- [x] Basic manual testing performed after deploying app

thorncorona committed:
Map viz done

thorncorona committed:
Map viz done

thorncorona committed:
Map viz done

thorncorona committed:
Finish geolocation

thorncorona committed:
Finish geolocation

thorncorona committed:
Finish geolocation

thorncorona committed:
Integrate geolocation into GeolocationDao

thorncorona committed:
Integrate geolocation into GeolocationDao

thorncorona committed:
Benchmark geoip lookup

thorncorona committed:
Benchmark geoip lookup

arunk054 commented on src/main/java/com/google/sps/Mock.java:
```diff
@@ -0,0 +1,28 @@
+
+import com.google.sps.models.PCAPAnalysis;
+
+public class Mock {
```
Should be more descriptive than just "Mock" and can you also clarify what "sps" stands for?

thorncorona replied:
Don't know what it stands for haha. It was the name of our package in the starter project. Changed package name

I changed it to `MockData`.

arunk054 commented on src/main/java/com/google/sps/Mock.java:
```diff
@@ -0,0 +1,28 @@
+
+public class Mock {
+
+  public static final String PCAP_ID = "8395f9e1-cc56-45d3-b801-aaa9a2ac8848";
```
Leave a comment on what this ID is used for. Is it a unique ID to the PCAP entry in the datastore?

thorncorona replied:
Just random data

arunk054 commented on src/main/java/com/google/sps/dao/GeolocationDao.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class GeolocationDao {
```
This is not really a DAO, you need to abstract it in a interface with getters, and you can extend it with a mock implementation. E.g. https://www.tutorialspoint.com/design_pattern/data_access_object_pattern.htm

arunk054 commented on src/main/java/com/google/sps/dao/GeolocationDao.java:
```diff
@@ -0,0 +1,24 @@
+
+public class GeolocationDao {
+
+  public static Map<String, String> ipCountryMap;
```
Avoid having static functions and entries, that is generally not a good practice.

thorncorona replied:
Fixed, integrated into DAO mock

arunk054 commented on src/main/java/com/google/sps/models/PCAPAnalysis.java:
```diff
@@ -0,0 +1,28 @@
+package com.google.sps.models;
+
+public class PCAPAnalysis {
```
if you can make this an interface it will be more extensible

thorncorona replied:
Fixed.

arunk054 commented on src/main/java/com/google/sps/models/PCAPAnalysis.java:
```diff
@@ -0,0 +1,28 @@
+    return id;
+  }
+
+  public void setId(String id) {
```
you should not have a setter for ID, as it should remain unchanged after object creation.

thorncorona replied:
Fixed

arunk054 commented on src/main/java/com/google/sps/models/PCAPAnalysis.java:
```diff
@@ -0,0 +1,28 @@
+  private String id;
+  private String[] ips;
+
+  public PCAPAnalysis(String id, String[] ips) {
```
Can you add a comment on what is the purpose of this class. What is it trying to achieve by encapsulating a id and set of IPs.

thorncorona replied:
Just a placeholder until I integrate w/ Eric.

arunk054 replied:
In that case add a comment and/or TODO saying it should be modified / removed after some integration.

thorncorona replied:
Changed.

arunk054 commented on src/main/java/com/google/sps/servlets/IPGeolocationServlet.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String id = getParameter(request, "PCAPId", "");
```
You could create a default PCAPID and store it as a constant. Your getParameter could return that instead. For now add a TODO may b until you add a default pcap id to your data store / mocks.

thorncorona replied:
Yep I changed the implementation in the new commits. For now I just set a default mock pcap

arunk054 commented on src/main/java/com/google/sps/servlets/IPGeolocationServlet.java:
```diff
@@ -0,0 +1,59 @@
+      Map<String, Integer> countryCount = new HashMap<>();
+
+      for (String ip: analysis.getIps()) {
+        String country = GeolocationDao.getCountryByIP(ip);
```
Are you taking care of case sensitive names in the DAO I believe. This assumes that the caller is sure the returned value is always case agnostic. Add a comment for that assumption.

thorncorona replied:
In new commits!

arunk054 commented on src/main/java/com/google/sps/servlets/IPGeolocationServlet.java:
```diff
@@ -0,0 +1,59 @@
+      countryCount.put("Jordan", 67);
+      countryCount.put("Estonia", 99);
+      countryCount.put("South Korea", 21);
+      countryCount.put("Canada", 66);
```
I believe all of this is going to be in your Mock DAO?

thorncorona replied:
Removed.

arunk054 commented on src/main/java/com/google/sps/servlets/IPGeolocationServlet.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      String id = getParameter(request, "PCAPId", "");
+      PCAPAnalysis analysis = AnalysisDao.getAnalysis(id);
```
You need to instantiate a DAO object in your init() method. That DAO instance can be a mock or can be talking to your data store or something else, based on the servlet config. Then in your servlet method here you will always call analysisDaoInstance.getAnalysis(id) irrespective of mocks or real. That will make it easier to inject mocks.

thorncorona replied:
In new commits

thorncorona committed:
Response to code review comments

- Add MaxMind CC-attribution text
- Fully implement DAOs
- Turn PCAP into interface
- Change package name to something more reasonable

thorncorona committed:
Response to code review comments

- Add MaxMind CC-attribution text
- Fully implement DAOs
- Turn PCAP into interface
- Change package name to something more reasonable

thorncorona committed:
Response to code review comments

- Add MaxMind CC-attribution text
- Fully implement DAOs
- Turn PCAP into interface
- Change package name to something more reasonable

thorncorona committed:
Add tests, fix init

thorncorona committed:
Add tests, fix init

thorncorona commented:
@arunk054 new changes pushed up!

thorncorona committed:
Remove extra mock stuff

thorncorona committed:
Remove extra mock stuff

arunk054 commented on scripts/BenchmarkGeoDB.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.List;
+import java.util.Random;
+
+public class BenchmarkGeoDB {
```
submit your benchmark stuff as a separate PR so it is easier for Philip to see.

thorncorona replied:
Removed from branch

arunk054 commented on src/main/java/com/google/netpcapanalysis/interfaces/models/PCAP.java:
```diff
@@ -0,0 +1,9 @@
+import java.net.InetAddress;
+import java.util.*;
+
+public interface PCAP {
```
PCAP has more to it than IP addresses. You should choose a different name for the interface. Melba has a PCAPData class that contains all the n tuples in a record, so a PCAP file is a list of PCAPData objects. So basically you should try to use that if you can, by creating a list of PCAP objects. You might only use the IP field that is fine and your mock would only mock IP.

thorncorona replied:
Can I use PCAP as a stub, and let Melba merge on top so we can use hers?

thorncorona replied:
@mnuzen Are you okay with merging your class into the interface?

arunk054 commented on src/main/java/com/google/netpcapanalysis/models/MockPCAP.java:
```diff
@@ -0,0 +1,51 @@
+    return ips;
+  }
+
+  public static InetAddress genIP() {
```
no reason this should be static. It can be a non-static method.

thorncorona replied:
Fixed.

arunk054 commented on src/main/java/com/google/netpcapanalysis/models/MockPCAP.java:
```diff
@@ -0,0 +1,51 @@
+        ip.isSiteLocalAddress()) {
+      try {
+        ip = InetAddress.getByName(
+            r.nextInt(256) + "." +
+                r.nextInt(256) + "." +
+                r.nextInt(256) + "." +
+                r.nextInt(256)
+        );
```
Did you do the rough math or a statistical computation on roughly how many iterations it takes before you get the IP you need. Seems a bit overkill for just mock data. You could as well run this offline in a separate script and then copy a set of valid IPs to a text file.

thorncorona replied:
I just copied it from the benchmark script. The time is pretty negligible. I used it to generate IPs for the benchmark script and 400K addresses didn't make the script much slower.

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java:
```diff
@@ -0,0 +1,39 @@
+
+  public GeolocationDaoImpl() {
+    try {
+      URL geoDBUrl = GeolocationDaoImpl.class.getClassLoader().getResource("GeoLite2-City.mmdb");
```
Declare the resource file name as a final static variable. Ideally we could read this from a properties file but that is future.

thorncorona replied:
Fixed

arunk054 reviewed (commented):
Looks much better. I think we need to work on abstracting the PCAP object out a little bit, coordinating with Melba's changes.

thorncorona committed:
Fix comments 1/2

thorncorona committed:
Fix comments 1/2

thorncorona committed:
Fix comments 1/2

thorncorona committed:
Remove extra files

thorncorona committed:
Remove extra files

thorncorona committed:
Merge branch 'map_viz' of github.com:mnuzen/step-capstone-2020 into map_viz

thorncorona committed:
Fix merge screwups

thorncorona commented:
Rebased.

thorncorona committed:
Merge branch 'master' into map_viz

thorncorona committed:
Merge branch 'master' into map_viz

thorncorona committed:
Integrated w/ datastore + tables

thorncorona commented:
@arunk054 I merged the PCAP models w/ Erik/Melba, can I get approved?

arunk054 commented on README.md:
```diff
@@ -4,3 +4,12 @@ This repo contains our pod's 2020 STEP Capstone project.
+## To use geolocation: 
+
+1. Download `GeoLite2-City.mmdb` from GDrive
+2. Put in repo under `data/GeoLite2-City.mmdb`
```
Try to use the .git folder for this or some other folder that you can mark in `.gitignore` so you don't accidentally commit this file.

thorncorona replied:
Yep in gitignore

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java:
```diff
@@ -0,0 +1,40 @@
+
+public class GeolocationDaoImpl implements GeolocationDao {
+
+  private static final String GEO_DB_LOCATION = "GeoLite2-City.mmdb";
```
Is this not under `data/` according to your README?

thorncorona replied:
It's mapped through Java resources.

arunk054 commented on src/main/java/com/google/netpcapanalysis/dao/GeolocationDaoImpl.java:
```diff
@@ -0,0 +1,40 @@
+  private File database;
+  private DatabaseReader reader;
+
+  public GeolocationDaoImpl() {
```
You could also have an overloaded constructor to pass in the GEO DB file path so that would make it a bit more flexible.

arunk054 commented on src/main/java/com/google/netpcapanalysis/mockdata/MockDataLoader.java:
```diff
@@ -1,33 +1,31 @@
 
   public void LoadData(ArrayList<PCAPdata> dataTable) { // uploads to datastore
-    GenericPCAPDao dataBase = new GenericPCAPDaoImpl();
+    PCAPDao dataBase = new PCAPDaoImpl();
```
You are creating the DAO every time instead of using a factory that maintains a global instance. So could lead to problems if two people trying to call this function on your website. The PCAPDaoImpl should be modified to a factory pattern like getInstance() instead allowing a constructor. May be do it on your next iteration.

thorncorona replied:
SG

arunk054 commented on src/main/tests/models/PCAPTest.java:
```diff
@@ -0,0 +1,5 @@
+package models;
+
+public class PCAPTest {
```
what is this class for btw?

thorncorona replied:
Scaffolding for Melba/Erik's stuff I think.

thorncorona committed:
Adjust readme
